libavfilter/sf_text2bitmap.c

```diff
@@ -0,0 +1,539 @@
+// Filter structures
+
+#define OFFSET(x) offsetof(ASS_Context, x)
+#define FLAGS (AV_OPT_FLAG_SUBTITLE_PARAM|AV_OPT_FLAG_FILTERING_PARAM)
+
+static const AVOption text2bitmap_options[] = {
+        { "w",          "output width",            OFFSET(width),      AV_OPT_TYPE_INT,    .min = 0, .max = 10000, .default_val.i64 = 0,    .flags = FLAGS },
+        { "width",      "output width",            OFFSET(width),      AV_OPT_TYPE_INT,    .min = 0, .max = 10000, .default_val.i64 = 0,    .flags = FLAGS },
+        { "h",          "output height",           OFFSET(height),     AV_OPT_TYPE_INT,    .min = 0, .max = 10000, .default_val.i64 = 0,    .flags = FLAGS },
+        { "height",     "output height",           OFFSET(height),     AV_OPT_TYPE_INT,    .min = 0, .max = 10000, .default_val.i64 = 0,    .flags = FLAGS },
+        { "n",          "number of output colors", OFFSET(num_colors), AV_OPT_TYPE_INT,    .min = 2, .max = 256,   .default_val.i64 = 256,  .flags = FLAGS },
+        { "num_colors", "number of output colors", OFFSET(num_colors), AV_OPT_TYPE_INT,    .min = 2, .max = 256,   .default_val.i64 = 256,  .flags = FLAGS },
+        { "f",          "fonts container",         OFFSET(fonts),      AV_OPT_TYPE_STRING,                         .default_val.str = NULL, .flags = FLAGS },
+        { "fonts",      "fonts container",         OFFSET(fonts),      AV_OPT_TYPE_STRING,                         .default_val.str = NULL, .flags = FLAGS },
+        { "fd",         "fonts directory",         OFFSET(fontsdir),   AV_OPT_TYPE_STRING,                         .default_val.str = NULL, .flags = FLAGS },
+        { "fontsdir",   "fonts directory",         OFFSET(fontsdir),   AV_OPT_TYPE_STRING,                         .default_val.str = NULL, .flags = FLAGS },
+        { .name =  NULL }
+};
+
+static const AVClass text2bitmap_class = {
+        .class_name       = "text2bitmap",
+        .item_name        = av_default_item_name,
+        .option           = text2bitmap_options,
+        .version          = LIBAVUTIL_VERSION_INT,
+        .category         = AV_CLASS_CATEGORY_FILTER,
+};
+
+static const AVFilterPad inputs[] = {
+    {
+        .name         = "default",
+        .type         = AVMEDIA_TYPE_SUBTITLE,
+        .filter_frame = filter_frame,
+        .config_props = config_input,
+    },
+};
+
+static const AVFilterPad outputs[] = {
+    {
+        .name         = "default",
+        .type         = AVMEDIA_TYPE_SUBTITLE,
+        .config_props = config_output,
+    },
+};
+
+const AVFilter ff_sf_text2bitmap = {
+        .name            = "text2bitmap",
+        .description     = NULL_IF_CONFIG_SMALL("Convert text subtitles to bitmap subtitles."),
+        .init            = init,
+        .uninit          = uninit,
+        .priv_size       = sizeof(ASS_Context),
+        .priv_class      = &text2bitmap_class,
+        FILTER_INPUTS(inputs),
+        FILTER_OUTPUTS(outputs),
+        FILTER_SUBFMTS(AV_SUBTITLE_FMT_TEXT, AV_SUBTITLE_FMT_ASS)
```

Comment:
You cannot use FILTER_SUBFMTS(). This is only for filters where input and output are of the same format.
You need to use the query_formats method instead.

Reply:
Can you explain what that function is supposed to do? I looked as some examples but I don't understand.

Reply:
This is about "format negotiation". A filter can support multiple combinations of input and output formats (not media types). The filtering implementation uses this to figure out a viable combination over a longer chain of filters.

In your case, you support only  AV_SUBTITLE_FMT_ASS and input and  AV_SUBTITLE_FMT_BITMAP as output.

You can look at the OCR filter, which is very similar just with the formats reversed.

At the bottom you need this:

https://github.com/softworkz/FFmpeg/blob/a66debd96ef9593a628ac7830819f345a98bdb02/libavfilter/sf_graphicsub2text.c#L1131

and then the same query_formats() function just with the formats reversed.